lib/plugins/cpuUsageThrottle.js

```diff
@@ -0,0 +1,139 @@
+/**
+ * cpuUsageThrottle
+ *
+ * Have restify reject a variable number of requests at the front door based on
+ * the process' current CPU usage. This plugin works by keeping a running
+ * average of the CPU usage over time and rejecting a percentage of requests
+ * based on the user specified CPU limit, maximum CPU limit, and the current
+ * value. This module will reject between 0% and 100% of all incomming requests
+ * when attempting to bring CPU usage back to a reasonable level. The method
+ * we use for calculating the CPU usage will work across platforms. This is
+ * an effective module for preventing high latencies caused by an over saturated
+ * restify server. This module is most effective when paired with an auto-scale
+ * policy at an orechestration layer.
+ *
+ * @param {Object} opts Configure this plugin.
+ * @param {Number} opts.limit The point at which restify will begin rejecting a
+ *    % of all requests at the front door. This value is a percentage like you
+ *    would see when running `top` on linux. For example .8 === 80%.
+ * @param {Number} opts.max The point at which restify will reject 100% of all
+ *    requests at the front door. This is used in conjunction with limit to
+ *    determine what % of traffic restify needs to reject when attempting to
+ *    bring the average load back within tolerable thresholds. Since Node.js is
+ *    single threaded, the default for this is 1. In some rare cases, a Node.js
+ *    process can exceed 100% CPU usage and you will want to update this value.
+ * @param {Number} opts.interval How frequently to check if we should be
+ *    rejecting/accepting connections. It's best to keep this value as low as
+ *    possible without creating a significant impact on perfomance in order to
+ *    keep your server as responsive to change as possible.
+ * @param {Number} opts.halfLife When we sample the CPU usage on an interval,
+ *    we create a series of data points. We take these points and calculate a
+ *    moving average. The halfLife indicates how quickly a point "decays" to
+ *    half it's value in the moving average. The lower the halfLife, the more
+ *    impact newer data points have on the average. If you want to be extremely
+ *    responsive to spikes in CPU usage, set this to a lower value. If you want
+ *    your process to put more emphasis on recent historical CPU usage when
+ *    determininng whether it should shed load, set this to a higher value. The
+ *    unit is in ms. The default is to set this to the same value as interval.
+ * @param {Error} [opts.err] A restify error used as a response when the
+ *    cpu usage limit is exceeded
+ * @returns {Function} middleware to be registered on server.pre
+ */
+function cpuUsageThrottle (opts) {
+
+    // Scrub input and populate our configuration
+    assert.object(opts, 'opts');
+    assert.number(opts.limit, 'opts.limit');
+    assert.optionalNumber(opts.max, 'opts.max');
+    assert.number(opts.interval, 'opts.interval');
```

Comment:
Is it possible to have a sane default for this? Just wondering we propose folks to figure out what this value should be.

Reply:
Probably right... I keep coming back to 500ms, so recheck every 1/2 second. Though I'm not sure if that is aggressive enough.

Reply:
Since this is a super low overhead function and results in reading from a single file descriptor, I'm going to have it default to checking every 1/4 of a second.